src/common/greptimedb-telemetry/src/lib.rs

```diff
@@ -0,0 +1,371 @@
+            GreptimeDBTelemetryTask::Disable => Ok(()),
+        }
+    }
+
+    pub async fn stop(&self) -> Result<()> {
+        match self {
+            GreptimeDBTelemetryTask::Enable(task) => task.stop().await,
+            GreptimeDBTelemetryTask::Disable => Ok(()),
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct StatisticData {
+    pub os: String,
+    pub version: String,
+    pub arch: String,
+    pub mode: Mode,
+    pub git_commit: String,
+    pub nodes: Option<i32>,
+    pub uuid: String,
+}
+
+#[derive(Serialize, Deserialize, Debug, Eq, PartialEq)]
+#[serde(rename_all = "lowercase")]
+pub enum Mode {
+    Distributed,
+    Standalone,
+}
+
+#[async_trait::async_trait]
+pub trait Collector {
+    fn get_version(&self) -> String {
+        env!("CARGO_PKG_VERSION").to_string()
+    }
+
+    fn get_git_hash(&self) -> String {
+        env!("GIT_COMMIT").to_string()
+    }
+
+    fn get_os(&self) -> String {
+        env::consts::OS.to_string()
+    }
+
+    fn get_arch(&self) -> String {
+        env::consts::ARCH.to_string()
+    }
+
+    fn get_mode(&self) -> Mode;
+
+    fn get_retry(&self) -> i32;
+
+    fn inc_retry(&mut self);
+
+    fn set_uuid_cache(&mut self, uuid: String);
+
+    fn get_uuid_cache(&self) -> Option<String>;
+
+    async fn get_nodes(&self) -> Option<i32>;
+
+    fn get_uuid(&mut self) -> Option<String> {
+        match self.get_uuid_cache() {
+            Some(uuid) => Some(uuid),
+            None => {
+                if self.get_retry() > 3 {
+                    return None;
+                }
+                match default_get_uuid() {
+                    Some(uuid) => {
+                        self.set_uuid_cache(uuid.clone());
+                        Some(uuid)
+                    }
+                    None => {
+                        self.inc_retry();
+                        None
+                    }
+                }
+            }
+        }
+    }
+}
+
+pub fn default_get_uuid() -> Option<String> {
+    let path = (*TELEMETRY_UUID_FILE_NAME).as_path();
+    match std::fs::read(path) {
+        Ok(bytes) => Some(String::from_utf8_lossy(&bytes).to_string()),
+        Err(e) => {
+            if e.kind() == ErrorKind::NotFound {
+                let uuid = uuid::Uuid::new_v4().to_string();
+                let _ = std::fs::write(path, uuid.as_bytes());
+                Some(uuid)
+            } else {
+                None
+            }
+        }
+    }
+}
+
+// Report version info to GreptimeDB.
+// We do not collect any identity-sensitive information.
+// This task is scheduled to run every 30 minutes.
+// The task will be disabled default. It can be enabled by setting the build feature `greptimedb-telemetry`
+// Collector is used to collect the version info. It can be implemented by different components.
+// client is used to send the HTTP request to GreptimeDB.
+// telemetry_url is the GreptimeDB url.
```

Comment:
```suggestion
/// Report version info to GreptimeDB.
/// We do not collect any identity-sensitive information.
/// This task is scheduled to run every 30 minutes.
/// The task will be disabled default. It can be enabled by setting the build feature `greptimedb-telemetry`
/// Collector is used to collect the version info. It can be implemented by different components.
/// client is used to send the HTTP request to GreptimeDB.
/// telemetry_url is the GreptimeDB url.
```